riscv: linker: try to use name of register

maass-hamburg opened the pull request:
Instead of always using the second register the SPI controller for memory mapping the jedec,spi-nor flash, use the name of the register. As a fallback, if the a register with that name doesn't exists use the second register.

To be able to add this functionality, this PR also adds DT_REG_HAS_NAME, DT_REG_ADDR_BY_NAME_OR, DT_REG_SIZE_BY_NAME_OR, DT_INST_REG_HAS_NAME, DT_INST_REG_SIZE_BY_NAME_OR and DT_INST_REG_ADDR_BY_NAME_OR. (That is also needed for https://github.com/zephyrproject-rtos/zephyr/pull/73446)

maass-hamburg committed:
dts: devicetree.h: add DT_REG_HAS_NAME

Add DT_REG_HAS_NAME,
DT_REG_ADDR_BY_NAME_OR,
DT_REG_SIZE_BY_NAME_OR,
DT_INST_REG_HAS_NAME,
DT_INST_REG_SIZE_BY_NAME_OR and
DT_INST_REG_ADDR_BY_NAME_OR.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
riscv: linker: try to use name of register

get the address of a memory mapped
jedec,spi-nor flash by name and as a fallback
use the second register.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

decsny reviewed (changes requested):
need to add test cases for DT additions and update API version

maass-hamburg committed:
tests: devicetree: add tests for DT_REG_HAS_NAME

Add tests for
DT_REG_HAS_NAME,
DT_REG_ADDR_BY_NAME_OR,
DT_REG_SIZE_BY_NAME_OR,
DT_INST_REG_HAS_NAME,
DT_INST_REG_SIZE_BY_NAME_OR and
DT_INST_REG_ADDR_BY_NAME_OR.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

decsny commented on tests/lib/devicetree/api/src/main.c:
I think it would make sense to include lines to test the "OR" fallback case

maass-hamburg replied:
These are in the lines 572, 585, 624 and 637. Do you need more than that?

decsny replied:
I see, I read it wrong

maass-hamburg commented:
By the way if somebody has a better name for the register address than `flash_mmap`, I'd like to hear it.
Maybe the linux dts has something already, but there I couldn't find it.

maass-hamburg commented:
ping @galak as Assignee

57300 reviewed (approved):
> By the way if somebody has a better name for the register address than `flash_mmap`, I'd like to hear it. Maybe the linux dts has something already, but there I couldn't find it.

In Zephyr, the existing SPI controllers use the name `mem` for this (while the main register block is called `control`).

Linux bindings have used a bunch of different names: `memory`, `fspi_mmap`, `qspi_mmap`, `qspi_mm`, `QuadSPI-memory`, and maybe more. So, there's not much standardization there at least.
